File: fragments/api/fastify/blockchain/solana_balance.ts
import { type FastifyInstance, type FastifyPluginOptions } from "fastify";
import { isAddress } from "@solana/kit";
import { getBalance } from "../../../solana_balance/solana_balance_utils";

export const routes = (fastify: FastifyInstance, _: FastifyPluginOptions) => {
  fastify.get<{
    Params: { address: string };
    Reply: {
      200: { balance: string };
      400: { error: string };
      500: { error: string };
    };
  }>("/balance/:address", async (request, reply) => {
    try {
      const { address } = request.params;

      if (!isAddress(address)) {
        return reply.code(400).send({ error: "Invalid Solana address" });
      }

      const balance = await getBalance(address);
      return {
        balance: balance.toString(),
      };
    } catch (error) {
      request.log.error(error, "Error fetching balance");
      return reply.code(500).send({ error: "Internal Server Error" });
    }
  });
};

File: fragments/apis/fastapi/blockchain/solana_balance.py
import logging
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from solders.pubkey import Pubkey

from fragments.solana_balance.solana_balance_utils import get_balance

solana_balance_router = APIRouter()
logger = logging.getLogger(__name__)


@solana_balance_router.get("/{address}", status_code=200)
async def get_solana_balance(address: str):
    try:
        pubkey = Pubkey.from_string(address)
    except ValueError:
        return JSONResponse(
            status_code=400,
            content={"error": "Invalid Solana address"},
        )

    try:
        balance = await get_balance(pubkey)
        return {"balance": str(balance)}
    except Exception as e:
        logger.error("Error fetching balance: %s", e)
        raise HTTPException(status_code=500, detail="Internal Server Error") from e
